File: app/src/commands/battle.ts
// Copyright (C) Call of Nil contributors
// SPDX-License-Identifier: AGPL-3.0-only

import { clamp } from 'es-toolkit';
import { toU8 } from '@/lib/number';
import { invoke } from '@tauri-apps/api/core';

export async function simulateBattle(args: {
  attacker?: Option<readonly Squad[]>;
  defender?: Option<readonly Squad[]>;
  wall?: Option<BuildingLevel>;
}) {
  args.attacker ??= [];
  args.defender ??= [];
  args.wall ??= 0;

  const stats = NIL.world.getStats();
  if (stats) {
    const minWall = stats.getBuildingMinLevel('wall');
    const maxWall = stats.getBuildingMaxLevel('wall');
    if (typeof minWall === 'number' && typeof maxWall === 'number') {
      args.wall = clamp(toU8(args.wall), minWall, maxWall);
    }
  }

  return invoke<BattleResult>('simulate_battle', { req: args });
}

File: crates/nil-core/src/world/battle.rs
// Copyright (C) Call of Nil contributors
// SPDX-License-Identifier: AGPL-3.0-only

use crate::battle::{Battle, BattleResult};
use crate::error::Result;
use crate::infrastructure::prelude::BuildingLevel;
use crate::military::squad::Squad;
use crate::world::World;

impl World {
  pub fn simulate_battle(
    &self,
    attacker: &[Squad],
    defender: &[Squad],
    wall: BuildingLevel,
  ) -> Result<BattleResult> {
    let wall_stats = if wall > BuildingLevel::ZERO {
      self
        .stats
        .infrastructure
        .wall()
        .get(wall)
        .map(Some)?
    } else {
      None
    };

    let result = Battle::builder()
      .attacker(attacker)
      .defender(defender)
      .maybe_wall(wall_stats)
      .build()
      .battle_result();

    Ok(result)
  }
}
